File: typescript/src/main/vsdk/toolkit/app/options/OptionsGroupCore.ts
import { ColorRgb } from "../../common/ColorRgb";
import { RenderOptions } from "../../common/RenderOptions";
import { OptionBase } from "../../common/commandLineOptions/OptionBase";
import { OptionGroup } from "../../common/commandLineOptions/OptionGroup";
import { OptionParser } from "../../common/commandLineOptions/OptionParser";
import { TypedOption } from "../../common/commandLineOptions/TypedOption";
import { ParseRuntimeContext } from "../../io/context/ParseRuntimeContext";
import { Background } from "../../scene/Background";
import { Camera } from "../../scene/Camera";
import { ConstantColorBackground } from "../../scene/ConstantColorBackground";
import { Scene } from "../../scene/Scene";
import { ToneMappingContext } from "../../tonemap/ToneMappingContext";
import { EnumAppOptions } from "./EnumAppOptions";
import { EnumBackgroundMode } from "./EnumBackgroundMode";
import { OptionTextUtils } from "./OptionTextUtils";
import { OptionsGroupCamera } from "./OptionsGroupCamera";
import { OptionsGroupRender } from "./OptionsGroupRender";
import { OptionsGroupToneMapping } from "./OptionsGroupToneMapping";

export class OptionsGroupCore {
  private static readonly DEFAULT_NUMBER_OF_QUARTIC_DIVISIONS = 4;
  private static readonly DEFAULT_FORCE_ONE_SIDED = true;
  private static readonly DEFAULT_BACKGROUND_COLOR = new ColorRgb(0.0, 0.0, 0.0);

  private static numberOfQuarterCircleDivisions = OptionsGroupCore.DEFAULT_NUMBER_OF_QUARTIC_DIVISIONS;
  private static fileOptionsForceOneSidedSurfaces = 0;
  private static outputImageWidth = 1920;
  private static outputImageHeight = 1080;
  private static glutDebugEnabled = 0;
  private static backgroundMode = EnumBackgroundMode.NONE;
  private static backgroundColor = new ColorRgb(
    OptionsGroupCore.DEFAULT_BACKGROUND_COLOR.r,
    OptionsGroupCore.DEFAULT_BACKGROUND_COLOR.g,
    OptionsGroupCore.DEFAULT_BACKGROUND_COLOR.b
  );

  private constructor() {
  }

  public static parse(
    argc: number[],
    argv: string[],
    parseSession: ParseRuntimeContext,
    scene: Scene,
    renderOptions: RenderOptions,
    toneMapOptions: ToneMappingContext,
    imageOutputWidth: number[],
    imageOutputHeight: number[],
    glutDebugEnabledOut: boolean[],
    toneMapNameOut: string[]
  ): void {
    const oneSidedSurfaces = [parseSession.singleSided];
    const conicSubDivisions = [parseSession.numberOfQuarterCircleDivisions];

    OptionsGroupCore.commandLineGeneralProgramParseOptions(
      argc,
      argv,
      oneSidedSurfaces,
      conicSubDivisions,
      imageOutputWidth,
      imageOutputHeight,
      glutDebugEnabledOut
    );

    parseSession.singleSided = oneSidedSurfaces[0];
    parseSession.parserConfig.singleSided = oneSidedSurfaces[0];
    parseSession.numberOfQuarterCircleDivisions = conicSubDivisions[0];
    parseSession.parserConfig.numberOfQuarterCircleDivisions = conicSubDivisions[0];

    OptionsGroupRender.renderParseOptions(argc, argv, renderOptions);
    OptionsGroupToneMapping.toneMapParseOptions(argc, argv, toneMapNameOut, toneMapOptions);

    if (scene.camera === null) {
      scene.camera = new Camera();
    }
    OptionsGroupCamera.cameraParseOptions(
      argc,
      argv,
      scene.camera,
      imageOutputWidth[0],
      imageOutputHeight[0]
    );
  }

  public static createBackground(): Background | null {
    return OptionsGroupCore.commandLineCreateBackground();
  }

  public static commandLineCreateBackground(): Background | null {
    if (OptionsGroupCore.backgroundMode === EnumBackgroundMode.SOLID) {
      return new ConstantColorBackground(OptionsGroupCore.backgroundColor);
    }
    return null;
  }

  private static commandLineParseFloat(text: string | null, value: number[]): boolean {
    if (text === null || value === null || value.length === 0) {
      return false;
    }

    const parsed = Number.parseFloat(text);
    if (Number.isNaN(parsed)) {
      return false;
    }
    value[0] = parsed;
    return true;
  }

  private static commandLineParseBackgroundColor(rArg: string | null, gArg: string | null, bArg: string | null, color: ColorRgb): boolean {
    const red = [0.0];
    const green = [0.0];
    const blue = [0.0];

    if (
      !OptionsGroupCore.commandLineParseFloat(rArg, red)
      || !OptionsGroupCore.commandLineParseFloat(gArg, green)
      || !OptionsGroupCore.commandLineParseFloat(bArg, blue)
    ) {
      return false;
    }

    if (
      red[0] < 0.0 || red[0] > 1.0
      || green[0] < 0.0 || green[0] > 1.0
      || blue[0] < 0.0 || blue[0] > 1.0
    ) {
      return false;
    }

    color.set(red[0], green[0], blue[0]);
    return true;
  }

  private static commandLineParseBackgroundOption(argc: number[], argv: string[]): void {
    if (argc === null || argc.length === 0 || argv === null) {
      return;
    }

    const args = argv as Array<string | null>;
    let writeIndex = 0;
    let readIndex = 0;
    while (readIndex < argc[0]) {
      const argument = args[readIndex];
      if (argument === null || argument !== "-background") {
        args[writeIndex++] = args[readIndex++];
        continue;
      }

      if (readIndex + 1 >= argc[0]) {
        process.stderr.write("Option '-background' requires a mode. Supported mode: solid.\n");
        readIndex += 1;
        continue;
      }

      const mode = args[readIndex + 1];
      if (!OptionTextUtils.equalsIgnoreCase(mode, "solid")) {
        process.stderr.write(
          `Invalid background mode '${mode}'. Expected '-background solid <r> <g> <b>'.\n`
        );
        readIndex += 2;
        continue;
      }

      if (readIndex + 4 >= argc[0]) {
        process.stderr.write(
          "Option '-background solid' requires three values in range [0.0, 1.0].\n"
        );
        readIndex += 2;
        continue;
      }

      const parsedColor = new ColorRgb();
      if (
        !OptionsGroupCore.commandLineParseBackgroundColor(
          args[readIndex + 2],
          args[readIndex + 3],
          args[readIndex + 4],
          parsedColor
        )
      ) {
        process.stderr.write(
          "Invalid '-background solid' color. Use '-background solid <r> <g> <b>' with values in [0.0, 1.0].\n"
        );
      }
      else {
        OptionsGroupCore.backgroundMode = EnumBackgroundMode.SOLID;
        OptionsGroupCore.backgroundColor = parsedColor;
      }
      readIndex += 5;
    }

    while (writeIndex < argc[0]) {
      args[writeIndex++] = null;
    }
    argc[0] = writeIndex;
  }

  private static mainForceOneSidedOption(value: TypedOption.MutableValue<number>): void {
    OptionsGroupCore.fileOptionsForceOneSidedSurfaces = value.value;
  }

  private static mainMonochromeOption(value: TypedOption.MutableValue<number>): void {
    OptionsGroupCore.numberOfQuarterCircleDivisions = value.value;
  }

  private static setIntTrue(value: TypedOption.MutableValue<number>): void {
    value.value = 1;
  }

  public static commandLineGeneralProgramParseOptions(
    argc: number[],
    argv: string[],
    oneSidedSurfaces: boolean[],
    conicSubDivisions: number[],
    imageOutputWidth: number[],
    imageOutputHeight: number[],
    glutDebugEnabledOut: boolean[]
  ): void {
    const appOptions = new EnumAppOptions();
    appOptions.width = OptionsGroupCore.outputImageWidth;
    appOptions.height = OptionsGroupCore.outputImageHeight;
    appOptions.nqcdivs = OptionsGroupCore.numberOfQuarterCircleDivisions;
    appOptions.yesValue = 1;
    appOptions.noValue = 0;
    appOptions.debug = 0;

    const widthOpt = new TypedOption<number>(
      "-width",
      TypedOption.reference(() => appOptions.width, (v) => {
        appOptions.width = v;
      }),
      1,
      null,
      null
    );
    const heightOpt = new TypedOption<number>(
      "-height",
      TypedOption.reference(() => appOptions.height, (v) => {
        appOptions.height = v;
      }),
      1,
      null,
      null
    );
    const nqcdivsOpt = new TypedOption<number>(
      "-nqcdivs",
      TypedOption.reference(() => appOptions.nqcdivs, (v) => {
        appOptions.nqcdivs = v;
      }),
      1,
      null,
      null
    );
    const forceOneSidedOpt = new TypedOption<number>(
      "-force-onesided",
      TypedOption.reference(() => appOptions.yesValue, (v) => {
        appOptions.yesValue = v;
      }),
      0,
      OptionsGroupCore.mainForceOneSidedOption,
      null
    );
    const dontForceOneSidedOpt = new TypedOption<number>(
      "-dont-force-onesided",
      TypedOption.reference(() => appOptions.noValue, (v) => {
        appOptions.noValue = v;
      }),
      0,
      OptionsGroupCore.mainForceOneSidedOption,
      null
    );
    const monochromaticOpt = new TypedOption<number>(
      "-monochromatic",
      TypedOption.reference(() => appOptions.yesValue, (v) => {
        appOptions.yesValue = v;
      }),
      0,
      OptionsGroupCore.mainMonochromeOption,
      null
    );
    const glutDebugOpt = new TypedOption<number>(
      "-glutDebug",
      TypedOption.reference(() => appOptions.debug, (v) => {
        appOptions.debug = v;
      }),
      0,
      OptionsGroupCore.setIntTrue,
      null
    );
    const registry: OptionBase[] = [
      TypedOption.REGISTER_OPTION(widthOpt, 5),
      TypedOption.REGISTER_OPTION(heightOpt, 6),
      TypedOption.REGISTER_OPTION(nqcdivsOpt, 3),
      TypedOption.REGISTER_OPTION(forceOneSidedOpt, 10),
      TypedOption.REGISTER_OPTION(dontForceOneSidedOpt, 14),
      TypedOption.REGISTER_OPTION(monochromaticOpt, 5),
      TypedOption.REGISTER_OPTION(glutDebugOpt, 6),
    ];

    OptionsGroupCore.fileOptionsForceOneSidedSurfaces = OptionsGroupCore.DEFAULT_FORCE_ONE_SIDED ? 1 : 0;
    OptionsGroupCore.numberOfQuarterCircleDivisions = OptionsGroupCore.DEFAULT_NUMBER_OF_QUARTIC_DIVISIONS;
    OptionsGroupCore.backgroundMode = EnumBackgroundMode.NONE;
    OptionsGroupCore.backgroundColor = new ColorRgb(
      OptionsGroupCore.DEFAULT_BACKGROUND_COLOR.r,
      OptionsGroupCore.DEFAULT_BACKGROUND_COLOR.g,
      OptionsGroupCore.DEFAULT_BACKGROUND_COLOR.b
    );
    OptionsGroupCore.glutDebugEnabled = appOptions.debug;

    OptionsGroupCore.commandLineParseBackgroundOption(argc, argv);

    const generalGroups = [
      new OptionGroup("global", registry, 7),
    ];
    OptionParser.parse(argc, argv, generalGroups, 1);

    OptionsGroupCore.outputImageWidth = appOptions.width;
    OptionsGroupCore.outputImageHeight = appOptions.height;
    OptionsGroupCore.numberOfQuarterCircleDivisions = appOptions.nqcdivs;
    OptionsGroupCore.glutDebugEnabled = appOptions.debug;

    if (oneSidedSurfaces !== null && oneSidedSurfaces.length > 0) {
      oneSidedSurfaces[0] = OptionsGroupCore.fileOptionsForceOneSidedSurfaces !== 0;
    }
    if (conicSubDivisions !== null && conicSubDivisions.length > 0) {
      conicSubDivisions[0] = OptionsGroupCore.numberOfQuarterCircleDivisions;
    }
    if (imageOutputWidth !== null && imageOutputWidth.length > 0) {
      imageOutputWidth[0] = OptionsGroupCore.outputImageWidth;
    }
    if (imageOutputHeight !== null && imageOutputHeight.length > 0) {
      imageOutputHeight[0] = OptionsGroupCore.outputImageHeight;
    }
    if (glutDebugEnabledOut !== null && glutDebugEnabledOut.length > 0) {
      glutDebugEnabledOut[0] = OptionsGroupCore.glutDebugEnabled !== 0;
    }
  }
}

File: java/testsuite/ApplicationCases/RenderparkApplication/src/main/java/vsdk/toolkit/app/options/OptionsGroupCore.java
package vsdk.toolkit.app.options;

import vsdk.toolkit.common.color.ColorRgb;
import vsdk.toolkit.material.RenderOptions;
import vsdk.toolkit.common.commandLineOptions.OptionBase;
import vsdk.toolkit.common.commandLineOptions.OptionGroup;
import vsdk.toolkit.common.commandLineOptions.OptionParser;
import vsdk.toolkit.common.commandLineOptions.TypedOption;
import vsdk.toolkit.io.context.ParseRuntimeContext;
import vsdk.toolkit.scene.Background;
import vsdk.toolkit.scene.ConstantColorBackground;
import vsdk.toolkit.scene.Scene;
import vsdk.toolkit.tonemap.ToneMappingContext;

public final class OptionsGroupCore {
    private static final int DEFAULT_NUMBER_OF_QUARTIC_DIVISIONS = 4;
    private static final boolean DEFAULT_FORCE_ONE_SIDED = true;
    private static final ColorRgb DEFAULT_BACKGROUND_COLOR = new ColorRgb(0.0, 0.0, 0.0);

    private static int numberOfQuarterCircleDivisions = DEFAULT_NUMBER_OF_QUARTIC_DIVISIONS;
    private static int fileOptionsForceOneSidedSurfaces = 0;
    private static int outputImageWidth = 1920;
    private static int outputImageHeight = 1080;
    private static int glutDebugEnabled = 0;
    private static EnumBackgroundMode backgroundMode = EnumBackgroundMode.NONE;
    private static ColorRgb backgroundColor = new ColorRgb(
        DEFAULT_BACKGROUND_COLOR.getR(),
        DEFAULT_BACKGROUND_COLOR.getG(),
        DEFAULT_BACKGROUND_COLOR.getB());

    private OptionsGroupCore() {
    }

    public static void parse(
        int[] argc,
        String[] argv,
        ParseRuntimeContext parseSession,
        Scene scene,
        RenderOptions renderOptions,
        ToneMappingContext toneMapOptions,
        int[] imageOutputWidth,
        int[] imageOutputHeight,
        boolean[] glutDebugEnabledOut,
        String[] toneMapNameOut)
    {
        boolean[] oneSidedSurfaces = new boolean[] {parseSession.singleSided};
        int[] conicSubDivisions = new int[] {parseSession.numberOfQuarterCircleDivisions};

        OptionsGroupCore.commandLineGeneralProgramParseOptions(
            argc,
            argv,
            oneSidedSurfaces,
            conicSubDivisions,
            imageOutputWidth,
            imageOutputHeight,
            glutDebugEnabledOut);

        parseSession.singleSided = oneSidedSurfaces[0];
        parseSession.parserConfig.singleSided = oneSidedSurfaces[0];
        parseSession.numberOfQuarterCircleDivisions = conicSubDivisions[0];
        parseSession.parserConfig.numberOfQuarterCircleDivisions = conicSubDivisions[0];

        OptionsGroupRender.renderParseOptions(argc, argv, renderOptions);
        OptionsGroupToneMapping.toneMapParseOptions(argc, argv, toneMapNameOut, toneMapOptions);
        OptionsGroupCamera.cameraParseOptions(
            argc,
            argv,
            scene.camera,
            imageOutputWidth[0],
            imageOutputHeight[0]);
    }

    public static Background createBackground() {
        return OptionsGroupCore.commandLineCreateBackground();
    }

    public static Background commandLineCreateBackground() {
        if ( backgroundMode == EnumBackgroundMode.SOLID ) {
            return new ConstantColorBackground(backgroundColor);
        }
        return null;
    }

    private static boolean commandLineParseFloat(String text, float[] value) {
        if ( text == null || value == null || value.length == 0 ) {
            return false;
        }

        try {
            value[0] = Float.parseFloat(text);
            return true;
        }
        catch ( NumberFormatException e ) {
            return false;
        }
    }

    private static boolean commandLineParseBackgroundColor(String rArg, String gArg, String bArg, ColorRgb color) {
        float[] red = new float[1];
        float[] green = new float[1];
        float[] blue = new float[1];

        if ( !OptionsGroupCore.commandLineParseFloat(rArg, red)
             || !OptionsGroupCore.commandLineParseFloat(gArg, green)
             || !OptionsGroupCore.commandLineParseFloat(bArg, blue) ) {
            return false;
        }

        if ( red[0] < 0.0f || red[0] > 1.0f
             || green[0] < 0.0f || green[0] > 1.0f
             || blue[0] < 0.0f || blue[0] > 1.0f ) {
            return false;
        }

        color.set(red[0], green[0], blue[0]);
        return true;
    }

    private static void commandLineParseBackgroundOption(int[] argc, String[] argv) {
        if ( argc == null || argc.length == 0 || argv == null ) {
            return;
        }

        int writeIndex = 0;
        int readIndex = 0;
        while ( readIndex < argc[0] ) {
            String argument = argv[readIndex];
            if ( argument == null || !"-background".equals(argument) ) {
                argv[writeIndex++] = argv[readIndex++];
                continue;
            }

            if ( readIndex + 1 >= argc[0] ) {
                System.err.printf("Option '-background' requires a mode. Supported mode: solid.\n");
                readIndex += 1;
                continue;
            }

            String mode = argv[readIndex + 1];
            if ( !OptionTextUtils.equalsIgnoreCase(mode, "solid") ) {
                System.err.printf(
                    "Invalid background mode '%s'. Expected '-background solid <r> <g> <b>'.\n",
                    mode);
                readIndex += 2;
                continue;
            }

            if ( readIndex + 4 >= argc[0] ) {
                System.err.printf(
                    "Option '-background solid' requires three values in range [0.0, 1.0].\n");
                readIndex += 2;
                continue;
            }

            ColorRgb parsedColor = new ColorRgb();
            if ( !OptionsGroupCore.commandLineParseBackgroundColor(
                     argv[readIndex + 2],
                     argv[readIndex + 3],
                     argv[readIndex + 4],
                     parsedColor) ) {
                System.err.printf(
                    "Invalid '-background solid' color. Use '-background solid <r> <g> <b>' with values in [0.0, 1.0].\n");
            }
            else {
                backgroundMode = EnumBackgroundMode.SOLID;
                backgroundColor = parsedColor;
            }
            readIndex += 5;
        }

        while ( writeIndex < argc[0] ) {
            argv[writeIndex++] = null;
        }
        argc[0] = writeIndex;
    }

    private static void mainForceOneSidedOption(TypedOption.MutableValue<Integer> value) {
        fileOptionsForceOneSidedSurfaces = value.value;
    }

    private static void mainMonochromeOption(TypedOption.MutableValue<Integer> value) {
        numberOfQuarterCircleDivisions = value.value;
    }

    private static void setIntTrue(TypedOption.MutableValue<Integer> value) {
        value.value = 1;
    }

    public static void commandLineGeneralProgramParseOptions(
        int[] argc,
        String[] argv,
        boolean[] oneSidedSurfaces,
        int[] conicSubDivisions,
        int[] imageOutputWidth,
        int[] imageOutputHeight,
        boolean[] glutDebugEnabledOut)
    {
        EnumAppOptions appOptions = new EnumAppOptions();
        appOptions.width = outputImageWidth;
        appOptions.height = outputImageHeight;
        appOptions.nqcdivs = numberOfQuarterCircleDivisions;
        appOptions.yesValue = 1;
        appOptions.noValue = 0;
        appOptions.debug = 0;

        TypedOption<Integer> widthOpt = new TypedOption<>(
            "-width",
            TypedOption.reference(() -> appOptions.width, v -> appOptions.width = v),
            1,
            null,
            null);
        TypedOption<Integer> heightOpt = new TypedOption<>(
            "-height",
            TypedOption.reference(() -> appOptions.height, v -> appOptions.height = v),
            1,
            null,
            null);
        TypedOption<Integer> nqcdivsOpt = new TypedOption<>(
            "-nqcdivs",
            TypedOption.reference(() -> appOptions.nqcdivs, v -> appOptions.nqcdivs = v),
            1,
            null,
            null);
        TypedOption<Integer> forceOneSidedOpt = new TypedOption<>(
            "-force-onesided",
            TypedOption.reference(() -> appOptions.yesValue, v -> appOptions.yesValue = v),
            0,
            OptionsGroupCore::mainForceOneSidedOption,
            null);
        TypedOption<Integer> dontForceOneSidedOpt = new TypedOption<>(
            "-dont-force-onesided",
            TypedOption.reference(() -> appOptions.noValue, v -> appOptions.noValue = v),
            0,
            OptionsGroupCore::mainForceOneSidedOption,
            null);
        TypedOption<Integer> monochromaticOpt = new TypedOption<>(
            "-monochromatic",
            TypedOption.reference(() -> appOptions.yesValue, v -> appOptions.yesValue = v),
            0,
            OptionsGroupCore::mainMonochromeOption,
            null);
        TypedOption<Integer> glutDebugOpt = new TypedOption<>(
            "-glutDebug",
            TypedOption.reference(() -> appOptions.debug, v -> appOptions.debug = v),
            0,
            OptionsGroupCore::setIntTrue,
            null);
        OptionBase[] registry = new OptionBase[] {
            TypedOption.REGISTER_OPTION(widthOpt, 5),
            TypedOption.REGISTER_OPTION(heightOpt, 6),
            TypedOption.REGISTER_OPTION(nqcdivsOpt, 3),
            TypedOption.REGISTER_OPTION(forceOneSidedOpt, 10),
            TypedOption.REGISTER_OPTION(dontForceOneSidedOpt, 14),
            TypedOption.REGISTER_OPTION(monochromaticOpt, 5),
            TypedOption.REGISTER_OPTION(glutDebugOpt, 6)
        };

        fileOptionsForceOneSidedSurfaces = DEFAULT_FORCE_ONE_SIDED ? 1 : 0;
        numberOfQuarterCircleDivisions = DEFAULT_NUMBER_OF_QUARTIC_DIVISIONS;
        backgroundMode = EnumBackgroundMode.NONE;
        backgroundColor = new ColorRgb(
            DEFAULT_BACKGROUND_COLOR.getR(),
            DEFAULT_BACKGROUND_COLOR.getG(),
            DEFAULT_BACKGROUND_COLOR.getB());
        glutDebugEnabled = appOptions.debug;

        OptionsGroupCore.commandLineParseBackgroundOption(argc, argv);

        OptionGroup[] generalGroups = new OptionGroup[] {
            new OptionGroup("global", registry, 7)
        };
        OptionParser.parse(argc, argv, generalGroups, 1);

        outputImageWidth = appOptions.width;
        outputImageHeight = appOptions.height;
        numberOfQuarterCircleDivisions = appOptions.nqcdivs;
        glutDebugEnabled = appOptions.debug;

        if ( oneSidedSurfaces != null && oneSidedSurfaces.length > 0 ) {
            oneSidedSurfaces[0] = fileOptionsForceOneSidedSurfaces != 0;
        }
        if ( conicSubDivisions != null && conicSubDivisions.length > 0 ) {
            conicSubDivisions[0] = numberOfQuarterCircleDivisions;
        }
        if ( imageOutputWidth != null && imageOutputWidth.length > 0 ) {
            imageOutputWidth[0] = outputImageWidth;
        }
        if ( imageOutputHeight != null && imageOutputHeight.length > 0 ) {
            imageOutputHeight[0] = outputImageHeight;
        }
        if ( glutDebugEnabledOut != null && glutDebugEnabledOut.length > 0 ) {
            glutDebugEnabledOut[0] = glutDebugEnabled != 0;
        }

    }
}
